p2p/src/config.rs

```diff
@@ -15,6 +15,49 @@
 
 use serde::{Deserialize, Serialize};
 
+/// Multicast DNS configuration.
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(tag = "type")]
+pub enum MdnsConfig {
+    Enabled {
+        /// Interval (in milliseconds) at which to poll the network for new peers.
+        query_interval: u64,
+        /// Use IPv6 for multicast DNS
+        enable_ipv6: bool,
+    },
+    Disabled,
+}
+
+impl MdnsConfig {
+    pub fn new() -> Self {
+        MdnsConfig::Disabled
+    }
+
+    pub fn from_options(
+        enable_mdns: bool,
+        query_interval: Option<u64>,
+        enable_ipv6: Option<bool>,
+    ) -> Self {
+        if enable_mdns {
+            MdnsConfig::Enabled {
+                query_interval: query_interval.unwrap_or(0),
+                enable_ipv6: enable_ipv6.unwrap_or(false),
+            }
+        } else {
+            assert!(
+                query_interval.is_none(),
```

Comment:
I wonder whether this can be done through a macro on the struct/enum somehow. Maybe throw in a TODO there describing that these checks shouldn't be done manually.